### ISSUE TITLE
SIGINT with threading backend keeps workers running

### ISSUE THREAD

**Issue by tgpfeiffer** [joblib/joblib#1304]
I have a program that uses joblib like
```python
results = Parallel(n_jobs, backend="threading", verbose=1)(
    delayed(f)(i, "joblib_threading") for i in range(n_jobs))
```
If the Python process running this code gets a SIGINT that is handled gracefully by the function containing the snippet above, the worker threads continue to run. That is, the main thread executes whatever comes in/after the `except` block, but the workers continue to perform their task.

I guess it is not a big problem for many users because normally you want Ctrl-C to terminate the Python process, which then also automatically stops all threads, but in some environments (e.g., Jupyter notebooks) all exceptions like KeyboardInterrupt and SystemExit are handled and presented to the user, so the process stays alive.

Is there any way we can finish the workers when a SIGINT is received? I guess it boils down to the old question of how to stop a thread in Python from the outside (most likely, checking a `threading.Event` periodically) but asking the users to implement something like that in their functions is maybe a bit tough, given that `joblib` provides an otherwise fairly simple to use and backend-agnostic interface.

Related: https://github.com/joblib/joblib/issues/181

**Comment by ogrisel** [comment]
Do all the thread receive the SIGINT or just one of them at random?

> If the Python process running this code gets a SIGINT that is handled gracefully by the function containing the snippet above

I am not sure what this means. Are signal handlers registered per-thread or per-process?

Could you give a minimal reproducer to clarify what you observe and what you would expect instead?

**Comment by tgpfeiffer** [comment]
> Do all the thread receive the SIGINT or just one of them at random?

Python programs always process signals in the main thread, see <https://docs.python.org/3/library/signal.html>:
"Python signal handlers are always executed in the main Python thread of the main interpreter, even if the signal was received in another thread. This means that signals can’t be used as a means of inter-thread communication. You can use the synchronization primitives from the [threading](https://docs.python.org/3/library/threading.html#module-threading) module instead."

> Could you give a minimal reproducer to clarify what you observe and what you would expect instead?

Sure, sorry.

```python
import sys, time
from joblib import Parallel, delayed

def f(job_id):
    for i in range(10):
        print(f"[{job_id}]: step {i}")
        time.sleep(1)
        
n_jobs = 4
try:
    results = Parallel(n_jobs, backend=sys.argv[1], verbose=1)(
        delayed(f)(i) for i in range(n_jobs))
except KeyboardInterrupt:
    print("got Ctrl-C")
    time.sleep(5)

print("sure everything is done now")
```

See how it behaves differently for the different backends:

```
jovyan@nb-7oq0er0449jm1dwp:~$ python example.py loky
[Parallel(n_jobs=4)]: Using backend LokyBackend with 4 concurrent workers.
[0]: step 0
[1]: step 0
[2]: step 0
[3]: step 0
[0]: step 1
[1]: step 1
[2]: step 1
[3]: step 1
[0]: step 2
[1]: step 2
[2]: step 2
[3]: step 2
^C[Parallel(n_jobs=4)]: Done   2 out of   4 | elapsed:    3.8s remaining:    3.8s
got Ctrl-C  # sleeps 5 seconds
sure everything is done now

jovyan@nb-7oq0er0449jm1dwp:~$ python example.py multiprocessing
[Parallel(n_jobs=4)]: Using backend MultiprocessingBackend with 4 concurrent workers.
[0]: step 0
[2]: step 0
[3]: step 0
[1]: step 0
[0]: step 1
[3]: step 1
[2]: step 1
[1]: step 1
[0]: step 2
[1]: step 2
[2]: step 2
[3]: step 2
^Cgot Ctrl-C  # sleeps 5 seconds
sure everything is done now

jovyan@nb-7oq0er0449jm1dwp:~$ python example.py threading
[Parallel(n_jobs=4)]: Using backend ThreadingBackend with 4 concurrent workers.
[0]: step 0
[1]: step 0
[3]: step 0
[2]: step 0
[0]: step 1
[2]: step 1
[3]: step 1
[1]: step 1
[0]: step 2
[2]: step 2
[3]: step 2
[1]: step 2
^Cgot Ctrl-C  # sleeps 5 seconds, but other threads continue printing
[0]: step 3
[2]: step 3
[3]: step 3
[1]: step 3
[2]: step 4
[3]: step 4
[1]: step 4
[0]: step 4
[2]: step 5
[3]: step 5
[1]: step 5
[0]: step 5
[0]: step 6
[1]: step 6
[3]: step 6
[2]: step 6
[0]: step 7
[1]: step 7
[3]: step 7
[2]: step 7
sure everything is done now  # process exits, forcefully killing all threads
```

**Comment by ogrisel** [comment]
Thanks, this is much clearer. Indeed we need to adjust the threading back to stop processing new tasks when the main parallel object has been shutdown. Not sure why this is not already the case. Currently the `ThreadingBackend` is still using the old `ThreadPool` class. I believe we should try to migrate it to use `concurrent.futures.ThreadPoolExecutor` instead but I am not sure if this will automatically fix this problem or not.

**Comment by ogrisel** [comment]
> Yes, I think with threads it's hard to stop processing from the "outside" without tearing down the whole process

I agree we should not try to hard-kill a Python thread, it might corrupt the state of the Python interpreter.

Indeed an event is a good idea, if we want to wake-up threads that are waiting on the submission queue.

For the `ThreadPoolExecutor`, note that the `concurrent.futures` API exposes a `.cancel()` method on submitted future objects. This should make it possible to prevent executing tasks that are pending in the queue.

The currently running tasks should still be running uninterrupted but I think this is fine as long as no more than one task executed per thread after the `KeyboardInterrupt` exception has been handled.
